internal/cms/signed_test.go

```diff
@@ -0,0 +1,418 @@
+			// err = err , err == nil, false, want error == false
+			if testcase.wantErr != (err != nil) {
+				t.Errorf("ParseSignedData.Verify() error = %v, wantErr %v", err, testcase.wantErr)
+			}
+		})
+	}
+}
+
+func TestVerifySigner(t *testing.T) {
+	ctx := context.Background()
+	// parse signed data
+	sigBytes, err := os.ReadFile("testdata/TimeStampToken.p7s")
+	if err != nil {
+		t.Fatal("failed to read test signature:", err)
+	}
+	signed, err := ParseSignedData(sigBytes)
+	if err != nil {
+		t.Fatal("ParseSignedData() error =", err)
+	}
+
+	roots := x509.NewCertPool()
+	certLen := len(signed.Certificates)
+	if certLen > 0 {
+		roots.AddCert(signed.Certificates[certLen-1])
+	}
+	intermediates := x509.NewCertPool()
+	for _, cert := range signed.Certificates {
+		intermediates.AddCert(cert)
+	}
+	opts := x509.VerifyOptions{
+		Roots:         roots,
+		KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageTimeStamping},
+		CurrentTime:   time.Date(2024, 1, 9, 0, 0, 0, 0, time.UTC),
+		Intermediates: intermediates,
+	}
+
+	t.Run("valid user provided signing certificate", func(t *testing.T) {
+		// verify with no root CAs and should fail
+		_, err = signed.VerifySigner(ctx, &signed.SignerInfos[0], signed.Certificates[0], opts)
+		if err != nil {
+			t.Errorf("ParseSignedData.Verify() error = %v, want nil", err)
+		}
+	})
+
+	t.Run("invalid user provided signing certificate", func(t *testing.T) {
+		// verify with no root CAs and should fail
+		_, err = signed.VerifySigner(ctx, &signed.SignerInfos[0], signed.Certificates[1], opts)
+		if err == nil {
+			t.Errorf("ParseSignedData.Verify() error = %v, want error", err)
+		}
+	})
+
+	t.Run("signerInfo is nil", func(t *testing.T) {
+		_, err = signed.VerifySigner(ctx, nil, signed.Certificates[0], opts)
+		if err == nil {
+			t.Error("ParseSignedData.Verify() error = nil, want error")
+		}
+	})
+
+	t.Run("certificate is nil", func(t *testing.T) {
+		// verify with no root CAs and should fail
+		_, err = signed.VerifySigner(ctx, &signed.SignerInfos[0], nil, opts)
+		if err == nil {
+			t.Error("ParseSignedData.Verify() error = nil, want error")
+		}
+	})
+}
+
+func TestVerifyAttributes(t *testing.T) {
+	testData := []struct {
+		name     string
+		filePath string
+		wantErr  bool
+	}{
+		{
+			name:     "without content type",
+			filePath: "testdata/TimeStampTokenWithoutSignedAttributeContentType.p7s",
+			wantErr:  true,
+		},
+		{
+			name:     "with invalid content type",
+			filePath: "testdata/TimeStampTokenInvalidSignedAttributeContentType.p7s",
+			wantErr:  true,
+		},
+		{
+			name:     "without signed attributes digest",
+			filePath: "testdata/TimeStampTokenWithoutSignedAttributeDigest.p7s",
+			wantErr:  true,
+		},
+		{
+			name:     "with SHA1 hash",
+			filePath: "testdata/TimeStampTokenWithSignedAttributeSHA1.p7s",
+			wantErr:  true,
+		},
+		{
+			name:     "with invalid signing time",
+			filePath: "testdata/TimeStampTokenWithSigingTime.p7s",
```

Comment:
typo:
This file name has a typo (`SigingTime`). Since we expect an error in it, we can rename it to,
```suggestion
			filePath: "testdata/TimeStampTokenWithInvalidSigningTime.p7s",
```

Reply:
Done.